File: internal/domain/role.go
package domain

import (
	"encoding/json"
	"time"
)

// Role represents a user role with permissions
type Role struct {
	ID          int64           `json:"id"`
	Name        string          `json:"name"`
	DisplayName string          `json:"display_name"`
	Description string          `json:"description,omitempty"`
	Permissions json.RawMessage `json:"permissions"`
	IsSystem    bool            `json:"is_system"`

	// Timestamps
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// TableName returns the database table name
func (Role) TableName() string {
	return "roles"
}

// GetPermissions returns the permissions as a string slice
func (r *Role) GetPermissions() ([]string, error) {
	var permissions []string
	if err := json.Unmarshal(r.Permissions, &permissions); err != nil {
		return nil, err
	}
	return permissions, nil
}

// HasPermission checks if the role has a specific permission
func (r *Role) HasPermission(permission string) bool {
	permissions, err := r.GetPermissions()
	if err != nil {
		return false
	}

	for _, p := range permissions {
		if p == "*" || p == permission {
			return true
		}
		// Check wildcard permissions (e.g., "customers.*" matches "customers.read")
		if len(p) > 2 && p[len(p)-2:] == ".*" {
			prefix := p[:len(p)-2]
			if len(permission) > len(prefix) && permission[:len(prefix)] == prefix {
				return true
			}
		}
	}
	return false
}

// Predefined role names
const (
	RoleSuperAdmin = "super_admin"
	RoleAdmin      = "admin"
	RoleManager    = "manager"
	RoleCashier    = "cashier"
	RoleSeller     = "seller"
)

File: web/src/types/role.ts
// Role types - mirrors internal/domain/role.go

export interface Role {
  id: number
  name: string
  display_name: string
  description?: string
  permissions: string[]
  is_system: boolean
  created_at: string
  updated_at: string
}

export interface CreateRoleInput {
  name: string
  display_name: string
  description?: string
  permissions: string[]
}

export interface UpdateRoleInput {
  name?: string
  display_name?: string
  description?: string
  permissions?: string[]
}

// Predefined role names
export const ROLE_SUPER_ADMIN = 'super_admin'
export const ROLE_ADMIN = 'admin'
export const ROLE_MANAGER = 'manager'
export const ROLE_CASHIER = 'cashier'
export const ROLE_SELLER = 'seller'

// Permission groups
export const PERMISSION_GROUPS = {
  customers: ['customers.read', 'customers.create', 'customers.update', 'customers.delete'],
  items: ['items.read', 'items.create', 'items.update', 'items.delete'],
  loans: ['loans.read', 'loans.create', 'loans.update', 'loans.delete'],
  payments: ['payments.read', 'payments.create', 'payments.update', 'payments.delete'],
  sales: ['sales.read', 'sales.create', 'sales.update', 'sales.delete'],
  cash: ['cash.read', 'cash.create', 'cash.update', 'cash.delete'],
  reports: ['reports.read', 'reports.export'],
  users: ['users.read', 'users.create', 'users.update', 'users.delete'],
  branches: ['branches.read', 'branches.create', 'branches.update', 'branches.delete'],
  categories: ['categories.read', 'categories.create', 'categories.update', 'categories.delete'],
  roles: ['roles.read', 'roles.create', 'roles.update', 'roles.delete'],
  settings: ['settings.read', 'settings.update'],
  audit: ['audit.read'],
  notifications: ['notifications.read', 'notifications.create', 'notifications.manage'],
  expenses: ['expenses.read', 'expenses.create', 'expenses.update', 'expenses.delete', 'expenses.approve'],
  transfers: ['transfers.read', 'transfers.create', 'transfers.approve', 'transfers.ship', 'transfers.receive', 'transfers.cancel'],
}

// Check if user has permission (with wildcard support)
export function hasPermission(userPermissions: string[], permission: string): boolean {
  for (const p of userPermissions) {
    if (p === '*' || p === permission) {
      return true
    }
    // Check wildcard permissions (e.g., "customers.*" matches "customers.read")
    if (p.endsWith('.*')) {
      const prefix = p.slice(0, -2)
      if (permission.startsWith(prefix)) {
        return true
      }
    }
  }
  return false
}
